<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 4 Sprint 2*

# Sprint Challenge - Neural Network Foundations

Table of Problems

1. [Defining Neural Networks](#Q1)
2. [Chocolate Gummy Bears](#Q2)
    - Perceptron
    - Multilayer Perceptron
4. [Keras MMP](#Q3)

<a id="Q1"></a>
## 1. Define the following terms:

- **Neuron:**
a function that can revieve and send information. When working with other neurons, they can make predictions

- **Input Layer:**
The first layer of a neural network where the flow of data begins and is feed forward for processing by other layers.

- **Hidden Layer:**
an intermediate layer that calculates weighted sums and is feed forward towards the output.
- **Output Layer:**
After the data is processed, this layer calculates the final results.
- **Activation:**
A function that decides how much a neuron contributes to the network based on its weighted sum.
- **Backpropagation:**
An algorithm that determines what weights were most responsible for causing som error in the output and uses gradient descent or some other type of optimization function to minimize the error.

## 2. Chocolate Gummy Bears <a id="Q2"></a>

Right now, you're probably thinking, "yuck, who the hell would eat that?". Great question. Your candy company wants to know too. And you thought I was kidding about the [Chocolate Gummy Bears](https://nuts.com/chocolatessweets/gummies/gummy-bears/milk-gummy-bears.html?utm_source=google&utm_medium=cpc&adpos=1o1&gclid=Cj0KCQjwrfvsBRD7ARIsAKuDvMOZrysDku3jGuWaDqf9TrV3x5JLXt1eqnVhN0KM6fMcbA1nod3h8AwaAvWwEALw_wcB). 

Let's assume that a candy company has gone out and collected information on the types of Halloween candy kids ate. Our candy company wants to predict the eating behavior of witches, warlocks, and ghosts -- aka costumed kids. They shared a sample dataset with us. Each row represents a piece of candy that a costumed child was presented with during "trick" or "treat". We know if the candy was `chocolate` (or not chocolate) or `gummy` (or not gummy). Your goal is to predict if the costumed kid `ate` the piece of candy. 

If both chocolate and gummy equal one, you've got a chocolate gummy bear on your hands!?!?!
![Chocolate Gummy Bear](https://ed910ae2d60f0d25bcb8-80550f96b5feb12604f4f720bfefb46d.ssl.cf1.rackcdn.com/3fb630c04435b7b5-2leZuM7_-zoom.jpg)

In [26]:
import pandas as pd
import numpy as np
np.random.seed(812)

candy = pd.read_csv('chocolate_gummy_bears.csv')

In [67]:
candy.head()

chocolate  gummy  ate
0          0      1    1
1          1      0    1
2          0      1    1
3          0      0    0
4          1      1    0

### Perceptron

To make predictions on the `candy` dataframe. Build and train a Perceptron using numpy. Your target column is `ate` and your features: `chocolate` and `gummy`. Do not do any feature engineering. :P

Once you've trained your model, report your accuracy. You will not be able to achieve more than ~50% with the simple perceptron. Explain why you could not achieve a higher accuracy with the *simple perceptron* architecture, because it's possible to achieve ~95% accuracy on this dataset. Provide your answer in markdown (and *optional* data anlysis code) after your perceptron implementation. 

In [32]:
# Start your candy perceptron here

X = candy[['chocolate', 'gummy']].values
y = candy['ate'].values

In [33]:
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derv(x):
    sx = sigmoid(x)
    return sx * (1 - sx)

In [34]:
# 2x1 matrix for weights since there are 2 inputs and one output
weights = 2 * np.random.random((2,1)) - 1

In [35]:
%%time
for _ in range(10):
    #weighted sum
    weighted_sum = np.dot(X,weights)
    
    #activate 
    activate = sigmoid(weighted_sum)
    
    #calculate error
    error = y - activate
    
    #adjust weights
    adj = error * sigmoid_derv(activate)
    
    # update weights 
    weights = weights + np.dot(X.T,adj)

print(f"Output: \n {activate}")

    
    

Output: 
 [[1. 1. 1. ... 1. 1. 1.]
 [1. 1. 1. ... 1. 1. 1.]
 [1. 1. 1. ... 1. 1. 1.]
 ...
 [1. 1. 1. ... 1. 1. 1.]
 [1. 1. 1. ... 1. 1. 1.]
 [1. 1. 1. ... 1. 1. 1.]]
Wall time: 54 s


In [36]:
print(f"accuracy: \n {1-(np.mean(np.square(y-activate)))}") # mean squared error

accuracy: 
 0.9370499999999999


### Multilayer Perceptron <a id="Q3"></a>

Using the sample candy dataset, implement a Neural Network Multilayer Perceptron class that uses backpropagation to update the network's weights. Your Multilayer Perceptron should be implemented in Numpy. 
Your network must have one hidden layer.

Once you've trained your model, report your accuracy. Explain why your MLP's performance is considerably better than your simple perceptron's on the candy dataset. 

In [64]:
class NeuralNetwork:
    def __init__(self):
        self.inputs = 2
        self.hidden = 4
        self.output = 1
        
        self.weight1 = np.random.rand(self.inputs,self.hidden)
        self.weight2 = np.random.rand(self.hidden,self.output)
        
    def sigmoid(self, s):
        return 1 / (1+np.exp(-s))
    
    def sigprime(self, s):
        return s * (1-s)
    
    def fforward(self, X):
        """
        Calculate the NN inference using feed forward.
        aka "predict"
        """
        
        # calculate the weighted sum of inputs => hidden layer
        # the dot product is established by multiplying each component
        # in one vector by each component in another vector, then add all the products
        # a = [2,3,4] 
        # b = [5,6,7]
        # dotproduct = (2*5) + (3*6) + (4*7) == 56
        self.hidden_sum = np.dot(X, self.weight1)
        
        # Activation of hidden sum
        self.hidden_activation = self.sigmoid(self.hidden_sum)
        
        # calculate the weighted sum of hidden layer => output
        self.output_sum = np.dot(self.hidden_activation,self.weight2)
        
        #final activation of output
        self.activated_output = self.sigmoid(self.output_sum)
        
        return self.activated_output
    
    
    def backprop(self,X,y,o):
        
        # calculate output error
        self.oerror = y - o
        
        # Apply Derivative of Sigmoid to error
        # How far off are we in relation to the Sigmoid f(x) of the output
        # ^- aka hidden => output
        self.odelta = self.oerror * self.sigprime(o)
        
        
        # z2 error 
        self.error_z2 = self.odelta.dot(self.weight2.T)
        # How much of that "far off" can explained by the input => hidden
        self.delta_z2 = self.error_z2 * self.sigprime(self.hidden_activation)
        
        # adjusting the weights from input => hidden
        self.weight1 = self.weight1 + X.T.dot(self.delta_z2)
        
        # adjusting the weights from hidden => output
        self.weight2 = self.weight2 + self.hidden_activation.T.dot(self.odelta)
        
        
    def train(self, X,y):
        o = self.fforward(X)
        print(f"shapes: O {o.shape},X: {X.shape}, Y: {y.shape}")
        self.backprop(X,y,o)

In [65]:
nn = NeuralNetwork()

for i in range(100):
    nn.train(X,y)

print(f"Actual: \n {y}")
print(f"Prediction: \n {nn.fforward(X)}")
print(f"accuracy: \n {1-(np.mean(np.square(y-nn.fforward(X))))}")


shapes: O (10000, 1),X: (10000, 2), Y: (10000,)


ValueError: shapes (10000,10000) and (1,4) not aligned: 10000 (dim 1) != 1 (dim 0)

P.S. Don't try candy gummy bears. They're disgusting. 

## 3. Keras MMP <a id="Q3"></a>

Implement a Multilayer Perceptron architecture of your choosing using the Keras library. Train your model and report its baseline accuracy. Then hyperparameter tune at least two parameters and report your model's accuracy.
Use the Heart Disease Dataset (binary classification)
Use an appropriate loss function for a binary classification task
Use an appropriate activation function on the final layer of your network.
Train your model using verbose output for ease of grading.
Use GridSearchCV or RandomSearchCV to hyperparameter tune your model. (for at least two hyperparameters)
When hyperparameter tuning, show you work by adding code cells for each new experiment.
Report the accuracy for each combination of hyperparameters as you test them so that we can easily see which resulted in the highest accuracy.
You must hyperparameter tune at least 3 parameters in order to get a 3 on this section.

In [2]:
import pandas as pd
from sklearn.preprocessing import StandardScaler

df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/heart.csv')
df = df.sample(frac=1)
print(df.shape)
df.head()

(303, 14)


age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
93    54    0   1       132   288    1        0      159      1      0.0   
181   65    0   0       150   225    0        0      114      0      1.0   
118   46    0   1       105   204    0        1      172      0      0.0   
124   39    0   2        94   199    0        1      179      0      0.0   
179   57    1   0       150   276    0        0      112      1      0.6   

     slope  ca  thal  target  
93       2   1     2       1  
181      1   3     3       0  
118      2   0     2       1  
124      2   0     2       1  
179      1   1     1       0

In [3]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [11]:
X = df[['age','sex','cp','trestbps','chol','fbs','restecg','thalach','exang','oldpeak','slope','ca','thal']].values
y = df['target'].values

(303, 13)

In [12]:
# create model
model = Sequential()
model.add(Dense(2,input_dim = 13,activation = 'relu'))
model.add(Dense(4,activation = 'sigmoid'))
model.add(Dense(1,activation = 'sigmoid'))
model.compile(loss = 'binary_crossentropy',optimizer = 'adam', metrics = ['accuracy'])

In [13]:
model.fit(X,y,epochs = 250, verbose = True)

Train on 303 samples
Epoch 1/250
303/303 [==============================] - 2s 5ms/sample - loss: 0.7246 - accuracy: 0.5446
Epoch 2/250
303/303 [==============================] - 0s 53us/sample - loss: 0.7206 - accuracy: 0.5446
Epoch 3/250
303/303 [==============================] - 0s 66us/sample - loss: 0.7168 - accuracy: 0.5446
Epoch 4/250
303/303 [==============================] - 0s 56us/sample - loss: 0.7143 - accuracy: 0.5446
Epoch 5/250
303/303 [==============================] - 0s 56us/sample - loss: 0.7119 - accuracy: 0.5446
Epoch 6/250
303/303 [==============================] - 0s 49us/sample - loss: 0.7098 - accuracy: 0.5446
Epoch 7/250
303/303 [==============================] - 0s 56us/sample - loss: 0.7079 - accuracy: 0.5446
Epoch 8/250
303/303 [==============================] - 0s 53us/sample - loss: 0.7059 - accuracy: 0.5446
Epoch 9/250
303/303 [==============================] - 0s 53us/sample - loss: 0.7044 - accuracy: 0.5446
Epoch 10/250
303/303 [======================

In [16]:
score = model.evaluate(X,y,verbose = False)
print(f"{model.metrics_names[1]}: {score[1]*100}")

accuracy: 54.45544719696045


# baseline accuracy of 54%

In [27]:
from sklearn.model_selection import GridSearchCV
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier


def create():
    # create model
    model = Sequential()
    model.add(Dense(2,input_dim = 13,activation = 'relu'))
    model.add(Dense(4,activation = 'sigmoid'))
    model.add(Dense(1,activation = 'sigmoid'))
    model.compile(loss = 'binary_crossentropy',optimizer = 'adam', metrics = ['accuracy'])
    return model
    
model2 = KerasClassifier(build_fn = create)

parameters = {'batch_size': [10,20,50,80,100],
             'epochs':[75]}

grid = GridSearchCV(estimator = model2, param_grid = parameters, n_jobs = -1)
results = grid.fit(X,y)

# report best 
print(f"Best: {results.best_score_} using {results.best_params_}")


Train on 303 samples
Epoch 1/75
303/303 [==============================] - 0s 1ms/sample - loss: 0.7445 - accuracy: 0.5446
Epoch 2/75
303/303 [==============================] - 0s 46us/sample - loss: 0.7394 - accuracy: 0.5446
Epoch 3/75
303/303 [==============================] - 0s 46us/sample - loss: 0.7359 - accuracy: 0.5446
Epoch 4/75
303/303 [==============================] - 0s 46us/sample - loss: 0.7327 - accuracy: 0.5446
Epoch 5/75
303/303 [==============================] - 0s 49us/sample - loss: 0.7295 - accuracy: 0.5446
Epoch 6/75
303/303 [==============================] - 0s 46us/sample - loss: 0.7263 - accuracy: 0.5446
Epoch 7/75
303/303 [==============================] - 0s 46us/sample - loss: 0.7235 - accuracy: 0.5446
Epoch 8/75
303/303 [==============================] - 0s 46us/sample - loss: 0.7208 - accuracy: 0.5446
Epoch 9/75
303/303 [==============================] - 0s 39us/sample - loss: 0.7181 - accuracy: 0.5446
Epoch 10/75
303/303 [==============================] 

# best with batch_size of 50 and 75 epochs
# accuracy 58%

In [28]:
model2 = KerasClassifier(build_fn = create)

parameters = {'batch_size': [50,80,100],
             'epochs':[200]}

grid = GridSearchCV(estimator = model2, param_grid = parameters, n_jobs = -1)
results = grid.fit(X,y)

# report best 
print(f"Best: {results.best_score_} using {results.best_params_}")

Train on 303 samples
Epoch 1/200
303/303 [==============================] - 1s 2ms/sample - loss: 0.6904 - accuracy: 0.5446
Epoch 2/200
303/303 [==============================] - 0s 39us/sample - loss: 0.6897 - accuracy: 0.5446
Epoch 3/200
303/303 [==============================] - 0s 36us/sample - loss: 0.6895 - accuracy: 0.5446
Epoch 4/200
303/303 [==============================] - 0s 39us/sample - loss: 0.6883 - accuracy: 0.5479
Epoch 5/200
303/303 [==============================] - 0s 36us/sample - loss: 0.6877 - accuracy: 0.5611
Epoch 6/200
303/303 [==============================] - 0s 46us/sample - loss: 0.6876 - accuracy: 0.5578
Epoch 7/200
303/303 [==============================] - 0s 39us/sample - loss: 0.6872 - accuracy: 0.5611
Epoch 8/200
303/303 [==============================] - 0s 39us/sample - loss: 0.6869 - accuracy: 0.5677
Epoch 9/200
303/303 [==============================] - 0s 43us/sample - loss: 0.6871 - accuracy: 0.5743
Epoch 10/200
303/303 [======================

# best using batch_size of 50 and 200 epochs
# accuracy 61%

In [29]:
# new model with best parameters and different optimizer
model = Sequential()
model.add(Dense(2,input_dim = 13,activation = 'sigmoid'))
model.add(Dense(5,activation = 'relu')) # added one more hidden layer
model.add(Dense(1,activation = 'softmax'))
model.compile(loss = 'binary_crossentropy',optimizer = 'sgd', metrics = ['accuracy'])

In [30]:
model.fit(X,y,epochs = 250,batch_size = 50, verbose = True)

Train on 303 samples
Epoch 1/250
303/303 [==============================] - 0s 1ms/sample - loss: 0.7859 - accuracy: 0.5446
Epoch 2/250
303/303 [==============================] - 0s 39us/sample - loss: 0.7834 - accuracy: 0.5446
Epoch 3/250
303/303 [==============================] - 0s 43us/sample - loss: 0.7808 - accuracy: 0.5446
Epoch 4/250
303/303 [==============================] - 0s 36us/sample - loss: 0.7785 - accuracy: 0.5446
Epoch 5/250
303/303 [==============================] - 0s 49us/sample - loss: 0.7763 - accuracy: 0.5446
Epoch 6/250
303/303 [==============================] - 0s 43us/sample - loss: 0.7743 - accuracy: 0.5446
Epoch 7/250
303/303 [==============================] - 0s 39us/sample - loss: 0.7718 - accuracy: 0.5446
Epoch 8/250
303/303 [==============================] - 0s 49us/sample - loss: 0.7696 - accuracy: 0.5446
Epoch 9/250
303/303 [==============================] - 0s 36us/sample - loss: 0.7647 - accuracy: 0.5446
Epoch 10/250
303/303 [======================

In [31]:
score = model.evaluate(X,y,verbose = False)
print(f"{model.metrics_names[1]}: {score[1]*100}")

accuracy: 66.66666865348816


# After tuning hyperparameters, was able to get up to 66% accuracy